translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,377 @@
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	translatorcommon "github.com/open-telemetry/opentelemetry-collector/translator/common"
+)
+
+const (
+	invalidMetricDescriptorType = ocmetrics.MetricDescriptor_Type(-1)
+)
+
+// ResourceMetricsToMetricsData converts metrics from OTLP to internal (OpenCensus) format
+func ResourceMetricsToMetricsData(resourceMetrics *otlpmetrics.ResourceMetrics) consumerdata.MetricsData {
+
+	node, resource := translatorcommon.ResourceToOC(resourceMetrics.Resource)
+	md := consumerdata.MetricsData{
+		Node:     node,
+		Resource: resource,
+	}
+
+	if len(resourceMetrics.Metrics) == 0 {
+		return md
+	}
+
+	metrics := make([]*ocmetrics.Metric, 0, len(resourceMetrics.Metrics))
+	for _, metric := range resourceMetrics.Metrics {
+		metrics = append(metrics, metricToOC(metric))
+	}
+
+	md.Metrics = metrics
+	return md
+}
+
+func metricToOC(metric *otlpmetrics.Metric) *ocmetrics.Metric {
+	labelKeys := allLabelKeys(metric)
+	return &ocmetrics.Metric{
+		MetricDescriptor: descriptorToOC(metric.MetricDescriptor, labelKeys),
+		Timeseries:       dataPointsToTimeseries(metric, labelKeys),
+		Resource:         nil,
+	}
+}
+
+func descriptorToOC(descriptor *otlpmetrics.MetricDescriptor, labelKeys []*ocmetrics.LabelKey) *ocmetrics.MetricDescriptor {
+	if descriptor == nil {
+		return nil
+	}
+
+	return &ocmetrics.MetricDescriptor{
+		Name:        descriptor.Name,
+		Description: descriptor.Description,
+		Unit:        descriptor.Unit,
+		Type:        descriptorTypeToOC(descriptor.Type),
+		LabelKeys:   labelKeys,
+	}
+}
+
+func allLabelKeys(metric *otlpmetrics.Metric) []*ocmetrics.LabelKey {
+	// NOTE: OpenTelemetry and OpenCensus have different representations of labels:
+	// - OC has a single "global" ordered list of label keys per metric in the MetricDescriptor;
+	// then, every data point has an ordered list of label values matching the key index.
+	// - In OTLP, every label is represented independently via StringKeyValue,
+	// i.e. theoretically points in the same metric may have different set of labels.
+	//
+	// So what we do in this translator:
+	// - Scan all points and their labels to generate a unique set of all label keys
+	// used across the metric, sort them and set in MetricDescriptor.
+	// - For each point we generate an ordered list of label values,
+	// matching the order of label keys returned here (see `labelsToOC` function).
+
+	// First, collect a set of unique keys
+	uniqueKeys := make(map[string]struct{}, 0)
+	for _, point := range metric.Int64Datapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+	for _, point := range metric.DoubleDatapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+	for _, point := range metric.HistogramDatapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+	for _, point := range metric.SummaryDatapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+
+	// Sort keys (this is actually optional, just for stable order)
```

Comment:
The comment is misplaced since it applies to line 114.
However, what's the benefit of sorting and having a stable order? Do we need it at all?

Reply:
This block is only needed for sorting, rephrased the comments and renamed variable to `sortedKeys` to make it clear.